### PR TITLE
Issue 174 acl

### LINKED ISSUE #174: User specific actions on profile page
Depending on users role and permission many button action changes on profile page. Will add more details on this

Implement manage user view, when the user clicks on manage button from colleague table show below view
image

All depending on user access different features and button changes
License admin see manage button
image

While view only access have view access

### PR DESCRIPTION
**Pull request by shehjadkhan-ansys**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #174 

Test URLs:
- Before: https://main--zemax--hlxsites.hlx.page/
- After: https://issue-174-acl--zemax--hlxsites.hlx.page/

### CODE REVIEW

File: blocks/user-licenses/user-actions.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  }
+}
+
+export async function addColleague(event, callback) {
+  event.preventDefault();
+  const formNode = event.target.parentNode;
+  const formData = new FormData(formNode);
+  formData.append('parentcustomerid_account', localStorage.getItem('parentcustomerid'));
+  const urlParams = new URLSearchParams(formData).toString();
+  const data = await execute('dynamics_add_colleague', `&${urlParams}`, 'POST');
```
Comment (synox) [inline blocks/user-licenses/user-actions.js:32]: When I commented on this earlier, I meant to pass the object like `{ parentcustomerid_account: localStorage.getItem('parentcustomerid') }` as a parameter to `execute`, so not every caller has to construct the URLSearchParams.

File: scripts/scripts.js [hunk 2]
```diff
@@ -130,6 +295,31 @@ function findReplaceJSON(jsonObj, data) {
   return jsonObj;
 }
 
+// Custom recursive deep copy function
+function deepCopy(obj) {
```
Comment (synox) [inline scripts/scripts.js:299]: Can probably be replaced with https://developer.mozilla.org/en-US/docs/Web/API/structuredClone

File: scripts/scripts.js [hunk 3]
```diff
@@ -71,6 +71,169 @@ export function createTag(tag, attributes, html) {
...
+  setTimeout(() => {
+    snackbar.classList.add('show');
+    setTimeout(() => {
+      snackbar.classList.remove('show');
+      snackbar.remove();
+    }, 2500);
+  }, 100);
+}
+
+export function createForm(config) {
```
Comment (synox) [inline scripts/scripts.js:90]: I recommend implementing the modal dialogs using https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

This makes compliance with accessibility much easier, and handling of backdrop clicks and escape key are much easier. Also this dialog is missing an option to close. 

Screenshot 2023-06-06 at 15 00 47

Comment (synox) [reply scripts/scripts.js:90]: Here is a simple dialog that I created: 

https://github.com/hlxsites/zemax/blob/83bef57fd5c80a247aef05a492087f763b9ddcac/blocks/header/header.js#L67-L79

How to open and close: 
https://github.com/hlxsites/zemax/blob/83bef57fd5c80a247aef05a492087f763b9ddcac/blocks/header/header.js#L341-L346

And how to handle click on backdrop: 
https://github.com/hlxsites/zemax/blob/83bef57fd5c80a247aef05a492087f763b9ddcac/blocks/header/header.js#L347-L354

And backdrop style: 
https://github.com/hlxsites/zemax/blob/83bef57fd5c80a247aef05a492087f763b9ddcac/blocks/header/header.css#L378-L380

**Review by synox — CHANGES_REQUESTED** [review]
I did not read all the code, but saw a few minor things that could be improved.

File: blocks/user-licenses/user-actions.js (hunk 1 above)
Comment (shehjadkhan-ansys) [reply blocks/user-licenses/user-actions.js]: Done

File: scripts/scripts.js (hunk 2 above)
Comment (shehjadkhan-ansys) [reply scripts/scripts.js:307]: I tried to use it it give me error not sure but looks like it is not handling method transfer properly
`function deepCopy(obj) {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }

  return structuredClone(obj);
}`
image

File: scripts/scripts.js (hunk 3 above)
Comment (shehjadkhan-ansys) [reply scripts/scripts.js:90]: I was able to do most of the part of this, for back drop click I would need some CSS help I did tried to incorporate what. I raised separate bug to handle that tomorrow. https://github.com/hlxsites/zemax/issues/197 
I would like to get this PR out soon so rest of the testing on profile page can continue

**Comment by shehjadkhan-ansys** [comment]
@synox Please take a look I have some pending issues which I will fix as a bug. thanks

File: scripts/scripts.js (hunk 3 above)
Comment (shehjadkhan-ansys) [reply scripts/scripts.js]: Some of the forms the one in screenshot are same in current production environment but I have fixed the close icon on top that will give the user option to close the modal

**Review by kmurugulla — CHANGES_REQUESTED** [review]

File: scripts/zemax-api.js [hunk 4]
```diff
@@ -1,7 +1,11 @@
 import { getEnvironmentConfig } from './zemax-config.js';
 
-export default async function execute(actionName, urlParams, methodType) {
+export default async function execute(actionName, urlConfig, methodType) {
   try {
+    let urlParams = urlConfig ? new URLSearchParams(urlConfig).toString() : '';
+    if (urlParams) {
+      urlParams = `&${urlParams}`;
+    }
     const userId = localStorage.getItem('auth0_id');
```
[the remark this answers is not in the document]
Comment (shehjadkhan-ansys) [reply scripts/zemax-api.js:9]: Sure null check can be added, this again ties up on our logic that profile page should not accessible to non logged in user, so ideally there should be any case of this method been called while we don't have auth0_id localstorage is not present

File: blocks/user-tickets/user-tickets.js [hunk 5]
```diff
@@ -1,153 +1,105 @@
...
-          data.forEach((ticket) => {
-            const trBody = document.createElement('tr');
-            const tdCaseNumber = document.createElement('td');
-            const caseLink = document.createElement('a');
-            caseLink.setAttribute('href', ticket.url);
-            const caseIdText = document.createTextNode(ticket.id);
-            caseLink.appendChild(caseIdText);
-            tdCaseNumber.appendChild(caseLink);
-            trBody.appendChild(tdCaseNumber);
+    renderUserTickets(userEmail, contactid, block);
```
[the remark this answers is not in the document]
Comment (shehjadkhan-ansys) [reply blocks/user-tickets/user-tickets.js:17]: I guess this is bigger question on profile page if somebreaks how we want to handle that and communicate to the user

File: scripts/scripts.js (hunk 2 above)
Comment (synox) [reply scripts/scripts.js:236]: Ah wait, structuredClone does not work with DOM elements. I assumed this are plain JS objects.

File: blocks/user-webroles/user-webroles.css [hunk 6]
```diff
@@ -54,4 +54,16 @@
...
     text-decoration: underline;
+}
+
+.user-webroles .more-info {
+    margin-left: 1.5rem;
+}
+
+a.more-info,
+a.visit-knowledgebase {
+    color: #0070c0!important;
```
[the remark this answers is not in the document]
Comment (shehjadkhan-ansys) [reply blocks/user-webroles/user-webroles.css]: Done

**Pull request closed without merging**